packages/router/router.ts

```diff
@@ -3038,12 +3087,129 @@ function shouldRevalidateLoader(
   return arg.defaultShouldRevalidate;
 }
 
+async function callRoutePipeline(
+  request: Request,
+  matches: AgnosticDataRouteMatch[],
+  requestContext: unknown,
+  handler:
+    | LoaderFunction
+    | ActionFunction
+    | LoaderFunctionWithMiddleware
+    | ActionFunctionWithMiddleware
+) {
+  // Avoid memory leaks since we don't control the key
+  let store = new WeakMap();
+  let middlewareContext: MiddlewareContext = {
+    get<T>(k: MiddlewareContextInstance<T>) {
+      if (store.has(k)) {
+        return store.get(k) as T;
+      }
+      return k.getDefaultValue();
+    },
+    set<T>(k: MiddlewareContextInstance<T>, v: T) {
+      if (typeof v === "undefined") {
+        throw new Error(
+          "You cannot set an undefined value in the middleware context"
+        );
+      }
+      store.set(k, v);
+    },
+    next: () => {},
+  };
```

Comment:
Create the `context` for this pipeline (scoped to a given loader/action execution)